src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -212,7 +212,8 @@ void CreatePrimaryRequest<I>::unlink_peer() {
       if (info->mirror_peer_uuids.size() == 0) {
         peer_uuid = peer;
         snap_id = snap_it.first;
-        break;
+        remove_snapshot(snap_id);
```

Comment:
I don't quite understand why you had to move `remove_snapshot` from `UnlinkPeerRequest` here? Couldn't `UnlinkPeerRequest` be fixed to do `remove_snapshot` when you wanted?

Reply:
Ah, never mind. It is because rbd-mirror also uses `UnlinkPeerRequest` and you don't want it to remove snapshots.

Reply:
> It is because rbd-mirror also uses `UnlinkPeerRequest` and you don't want it to remove snapshots.

The motivation is clear but I think we should leave `UnlinkPeerRequest::remove_snapshot()` in place, making the behavior conditional at `UnlinkPeerRequest` creation time.  The current approach where snapshots are removed only in `CreatePrimaryRequest` appears to be missing that `Mirror::peer_site_remove()` also invokes `UnlinkPeerRequest` with the expectation that snapshots would be removed.